### DISCUSSION TITLE
threading and schedulers

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#250] [Design & Development]
Like most of my recent discussion entries this one came from working to adapt pwmig to mspass.   It will take a bit to explain why the title I used has anything to do with that so hold on. 

Under the hood in pwmig is an old (and ugly) library called gclgrid.   The issue here comes up because of the algorithm gclgrid uses to implement a method it contains called "lookup".   The algorithm, as I recall from the paper we wrote more than 15 years is something called the "direction set method" (side issue is I've never been certain that is the right tag .   Point is a google search may not help you understand what it is. )   gclgrid was developed to handle distorted, regular grids.  that means the grid points map into arrays and can be indexed but the grid coordinates of every point can be anything as long as the grid doesn't fold on itself.  To be clear, a 3D scalar field has the following variables defined in the object definition (GCLscalarfield3d):
```
double ***x1,***x2,***x3;
double ***val;
```
For anyone not intimately familiar with C a symbol like *** once allocated can be referenced with three indices.  e.g. `x1[i][j][k]`. 

The lookup method, which is the topic here, finds the cell corresponding to a point (px1,px2,px3) inside the grid by an interative algorithm.  That iteration starts from a previously stored marking point defined as a private variable in the object.  For completeness, here is the private section declaration of GCLscalarfield (actually private member of parent called GCLgrid3d):
```
private:
	int ix1, ix2, ix3;
        bool fast_lookup;
```
THE ISSUE is that the lookup method REQUIRES AND ALTERS ix1, ix2, and ix3.   It does that for efficiency and stability.  In working through any 3d volume random access is rare.   Most algorithms work through the grid systematically along one set of generalized coordinates at a time or along a curve through the volume (ray tracing).   

Why this matters is that it dawns on me that this algorithm is anything but thread safe.  If two instances of a function calling lookup try to work on the same object at the same time they will fight each other.  In the best case it slows the code and in the worst case it could result in a runaway with dueling changes to ix1, ix2, and ix3.   

I can think of three solutions to this problem, none of which are very attractive and something we wouldn't want to advise a future collaborative developer to do without good reason:
1.  Redesign the algorithm.   That is not a fun proposition and one almost any outside person would likely balk at doing without very strong motivations.   For this case it might not be wrong because the existing library has some fundamental flaws anyway, but it would be very an annoying thing to have to undertake because a strength of the algorithm is that the lookup method is lightening fast compared to stateless lookup algorithms I've considered over the years.   
2. Use pthread mutex to keep a multithreaded implementation from altering the data at the same time.  There are a whole lot of reasons to not want to do that.  (a) I think the algorithm involved would not scale well if I had to lock the objects being referenced.  The reasons for that are deep in the code.  (b) A more complex locking hierarchy might improve performance, but debugging such would likely be a nightmare.   (c) A key question for this forum is I have no idea how dask and spark would interact with pthread calls inside a C++ function bound to python with pybind11?   i.e. this solution may not even be feasible.
3. It might be possible to redesign the object definitions to work around this problem.   There might be a solution for this particular algorithm where the data defined by ix1, ix2, and ix3 are passed as arguments instead of being held as a static member of the class.   The dark side of that is it requires a patch to an already messy library, but it is probably the one I'll at least start with. 

The bottom line issue for this discussion is this question raised above:   How do dask and spark interact with pthread mutex calls inside a C/C++ function?  Is that model even feasible?  Also how do we document and warn our evolving user base about the dangers of this kind of algorithm and what to do about it?
